src/platform/mt8196/lib/clk.c

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright(c) 2024 MediaTek. All rights reserved.
+ *
+ * Author: Hailong Fan <hailong.fan@mediatek.com>
+ */
+
+#include <platform/drivers/mt_reg_base.h>
+#include <rtos/clk.h>
+#include <rtos/wait.h>
+#include <sof/common.h>
+#include <sof/lib/cpu.h>
+#include <sof/lib/io.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/notifier.h>
+#include <sof/lib/uuid.h>
+#include <rtos/sof.h>
+#include <sof/trace/trace.h>
+
+SOF_DEFINE_REG_UUID(clkdrv_mt8196);
+
+DECLARE_TR_CTX(clkdrv_tr, SOF_UUID(clkdrv_mt8196_uuid), LOG_LEVEL_INFO);
+
+/* default voltage is 0.75V */
+const struct freq_table platform_cpu_freq[] = {
+	{  26000000, 26000},
+	{ 800000000, 26000},
+};
+
+STATIC_ASSERT(ARRAY_SIZE(platform_cpu_freq) == NUM_CPU_FREQ,
+	      invalid_number_of_cpu_frequencies);
+
+static SHARED_DATA struct clock_info platform_clocks_info[NUM_CLOCKS];
+
+static void clk_dsppll_enable(uint32_t value)
+{
+	tr_dbg(&clkdrv_tr, "clk_dsppll_enable %d\n", value);
+
+	switch (value) {
+	case ADSP_CLK_PLL_400M:
+		io_reg_write(MTK_ADSPPLL_CON1, MTK_PLL_DIV_RATIO_400M);
+		break;
+	case ADSP_CLK_PLL_800M:
+		io_reg_write(MTK_ADSPPLL_CON1, MTK_PLL_DIV_RATIO_800M);
+		break;
+	default:
+		tr_err(&clkdrv_tr, "invalid dsppll: %d\n", value);
+		return;
+	}
+
+	//todo
+	//io_reg_update_bits(MTK_ADSPPLL_CON3, MTK_PLL_PWR_ON, MTK_PLL_PWR_ON);
```

Comment:
Maybe clean this up, I guess these are now copied from platform to another as copies.